### DISCUSSION TITLE
Need help with generating Unique constraints in the declarative models.

### DISCUSSION THREAD

**Discussion by Vrushank-21** [agronholm/sqlacodegen#245] [Q&A]
```sql
CREATE TABLE "DUMMY" (
	"attr1"	integer NOT NULL,
	"attr2"	char NOT NULL,
	"attr3"	datetime NOT NULL,
	CONSTRAINT "pk1" PRIMARY KEY("attr1","attr3"),
	CONSTRAINT "uk" UNIQUE("attr1","attr2","attr3")
);
```

For the above table, what I want is
```python
class DUMMY(Base):
    __tablename__ = 'DUMMY'
    __table_args__ = (
        PrimaryKeyConstraint('attr3', name='pk1'),
        UniqueConstraint('attr1', 'attr2', 'attr3', name='uk')
    )

    attr1 = Column(Integer, nullable=False)
    attr2 = Column(CHAR, nullable=False)
    attr3 = Column(DateTime)
```

But this is what I get:
```python
class DUMMY(Base):
    __tablename__ = 'DUMMY'
    __table_args__ = (
        UniqueConstraint('attr1', 'attr2', 'attr3', name='uk'),
    )

    attr1 = Column(Integer, primary_key=True, nullable=False)
    attr2 = Column(CHAR, nullable=False)
    attr3 = Column(DateTime, primary_key=True, nullable=False)
```

**Comment by agronholm** [comment]
What's the problem with the generated code?

**Reply by agronholm** [reply]
Your SQL specifies a two-column primary key constraint, but your desired code only has a single column in the PK constraint. Why would you want sqlacodegen to generate the wrong code?

**Reply by Vrushank-21** [reply]
Is it possible to have both Primary and Unique constraint in __table_args__? That would be great for me

**Reply by agronholm** [reply]
Why? What's the difference? Anyway, SQLACodegen just generates the base model code for you; it's not expected to generate 1:1 the code the user wants. Just edit the resulting model file by hand.

**Reply by Vrushank-21** [reply]
The thing is I have raised this as an issue yesterday and It worked just fine for you. But when I run it, it works a little different. Can you please tell me the python, sqlalchemy and sqlite3 versions you are using?

**Reply by Vrushank-21** [reply]
> Why? What's the difference? Anyway, SQLACodegen just generates the base model code for you; it's not expected to generate 1:1 the code the user wants. Just edit the resulting model file by hand.

The thing is, there have been scenarios where only PrimaryKeyConstraint was generated and not the UniqueKey constraint. This is where I had issues

**Reply by agronholm** [reply]
The issue you raised yesteday was that it was not generating a unique constraint. Now you're complaining that it doesn't generate an explicit primary key constraint, but rather specifying `primary_key=True` in the columns, which is the same thing and which sqlacodegen has been doing from the beginning.

The subject of this discussion says you need help with generating Unique constraints, but then, according to the code you yourself posted, the unique constraint gets generated exactly like you wanted. The only difference is how the primary key is expressed. So what is it that you want?? As for my Python, sqlalchemy and sqlite3 versions, I already posted them yesterday. They haven't changed since.

**Reply by Vrushank-21** [reply]
> Your SQL specifies a two-column primary key constraint, but your desired code only has a single column in the PK constraint. Why would you want sqlacodegen to generate the wrong code?

Sincere apologies with the wrong DDL. The second primary key was an edit I made before copy-pasting here. The results I mentioned were for the case of 1 primary key only

**Reply by agronholm** [reply]
So what is it that you want then? Clearly the unique constraint is not the issue here (contrary to what the topic says).

**Reply by Vrushank-21** [reply]
Sorry about the bad explanation on my part. I honestly wish I could explain it properly.

So yesterday, when I had raised an issue, you asked me try running sqlacodegen on a test.sqlite3 file. It worked exactly the way I wanted. But when I used the same DDL to create the exact same database locally and ran sqlacodegen, the final code generated was slightly different from the one obtained from your file 'test.sqlite3'. And the result I obtained is the one I had mentioned above.

All I wanted to know was why this difference showed up and why was I not getting the same code you were getting.

**Reply by agronholm** [reply]
The only reason why sqlacodegen would generate an explicit `PrimaryKeyConstraint` would be if it had an explicit name that wasn't what would be generated by default. Does this answer your question?

**Reply by Vrushank-21** [reply]
Yes. Got it . Thank you.
